File: app/src/main/java/betrip/kitttn/architecturecomponentstest/vm/ActivityViewModel.java
package betrip.kitttn.architecturecomponentstest.vm;

import android.arch.lifecycle.ViewModel;

/**
 * @author kitttn
 */

public class ActivityViewModel extends ViewModel {
    public enum State {
        STARTED, SEARCH_LOADED
    }
    public State currentState = State.STARTED;

    public void openSearch() {
        currentState = State.SEARCH_LOADED;
    }

    public void closeApp() {
        currentState = State.STARTED;
    }
}

File: app/src/main/java/betrip/kitttn/architecturecomponentstest/vm/ActivityViewModel.kt
package betrip.kitttn.architecturecomponentstest.vm

import android.arch.lifecycle.ViewModel
import io.reactivex.processors.BehaviorProcessor

/**
 * @author kitttn
 */

sealed class NavigationState

class AppStartedState : NavigationState()

class SearchLoadedState : NavigationState()

data class CountryOpenedState(val countryName: String) : NavigationState()

class ActivityViewModel : ViewModel() {
    var currentState: NavigationState = AppStartedState()

    fun openSearch() {
        currentState = SearchLoadedState()
    }

    fun openCountryDetails(countryName: String) {
        currentState = CountryOpenedState(countryName)
    }

    fun closeApp() {
        currentState = AppStartedState()
    }
}